File: hyobin/week7/BOJ_1976.java
import java.io.*;
import java.util.*;
public class BOJ_1976 {
    static int n;
    static int m;
    static int[] parents;
    static int[] city;

    static int get(int a){
        if(parents[a] == a) return a;
        return parents[a] = get(parents[a]);
    }
    static boolean find(int a, int b){
        int aroot = get(a);
        int broot = get(b);

        if(aroot==broot) return true;
        return false;
    }

    static void union(int a, int b){
        int aroot = get(a);
        int broot = get(b);

        if(a<b){
            parents[broot] = aroot;
        }else {
            parents[aroot] = broot;
        }

    }

    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int m = Integer.parseInt(br.readLine());
        city = new int[m];
        StringTokenizer st;
        parents = new int[n+1];
        for (int i =1; i<=n ; i++){
            parents[i]=i;
        }
        for(int i=1; i<=n; i++){
            st = new StringTokenizer(br.readLine());
            for(int j=1; j<=n;j++){
                if(Integer.parseInt(st.nextToken()) == 1){
                    if(!find(i,j)) {
                        union(i, j);
                    }
                }
            }
        }
        st = new StringTokenizer(br.readLine());
        for(int i=0; i<m; i++){
            city[i] = Integer.parseInt(st.nextToken());
        }
        boolean isflag = false;
        for(int i=0; i<m-1; i++){
            if(!find(city[i], city[i+1])){

                isflag = true;
                break;
            }
        }
        if(isflag){
            System.out.println("NO");
        }else {
            System.out.println("YES");
        }

    }
}

File: SeongJae/week7/BOJ_1976.py
def findSet(a):
    if(parents[a]<0):
        return a
    parents[a] = findSet(parents[a])
    return parents[a]

def union(a,b):
    aRoot=findSet(a)
    bRoot=findSet(b)
    if(aRoot==bRoot):
        return False
    #집합 크기 관리
    #편의상 a집합에 b집합을 붙임
    parents[aRoot]+=parents[bRoot]
    parents[bRoot]=aRoot
    return True

n=int(input())
m=int(input())
parents=[-1]*(n+1)
arr=[list(map(int,input().split())) for _ in range(n)]
for i in range(1,n+1):
     for j in range(i,n+1):
         if arr[i-1][j-1]==1:
             union(i,j)
arr=list(map(int,input().split()))
isYes=True
for i in range(m-1):
     if(findSet(arr[i])==findSet(arr[i+1])):
          continue
     else:
          isYes=False
if isYes:
    print("YES")
else:
    print("NO")